python/pyarrow/_substrait.pyx

```diff
@@ -25,14 +26,84 @@ from pyarrow.includes.libarrow cimport *
 from pyarrow.includes.libarrow_substrait cimport *
 
 
-def run_query(plan):
+cdef shared_ptr[CTable] _process_named_table(dict named_args, const std_vector[c_string]& names):
+    cdef:
+        c_string c_name
+        shared_ptr[CTable] empty_table
+    py_names = []
+
+    # provider function could raise an exception
+    try:
+        for i in range(names.size()):
+            c_name = names[i]
+            py_names.append(frombytes(c_name))
+        return pyarrow_unwrap_table(named_args["provider"](py_names))
+    except Exception:
+        return empty_table
```

Comment:
I don't know that an empty table is the correct thing to return here.  Maybe we can return `CResult[shared_ptr[CTable]]` so that we can return an invalid status?  Ideally we would capture the exception message to put into the status too. I'm not sure if we have an example of that.

Reply:
How about doing a `raise` instead. 

```bash
pyarrow/tests/test_substrait.py::test_named_table_invalid_table_name
  /Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: 'pyarrow._substrait._process_named_table'
  
  Traceback (most recent call last):
    File "pyarrow/_substrait.pyx", line 39, in pyarrow._substrait._process_named_table
      return pyarrow_unwrap_table(named_args["provider"](py_names))
    File "/Users/vibhatha/github/fork/arrow/python/pyarrow/tests/test_substrait.py", line 223, in table_provider
      raise Exception("Unrecognized table name")
  Exception: Unrecognized table name
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/Users/vibhatha/github/fork/arrow/python/pyarrow/tests/test_substrait.py", line 252, in test_named_table_invalid_table_name
      substrait.run_query(buf, table_provider)
  ValueError: Error occurred in extracting NamedTable : Unrecognized table name
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
```

I want to fix this warning. Not sure if we can handle it here properly.

Reply:
`raise` looks good to me.  I had forgotten that we were inside SafeCallIntoPython which handles that for us.  Although in that case do we need to even catch the exception?

Reply:
I am not quite sure if there is a `SafeCallIntoPython`, aren't we just wraping `cython` function into `std::function`? 
If we don't handle it here, the `pytests` gives the aforementioned warning. What do you think?

Reply:
BindFunction used SafeCallIntoPython.  I  think what you suggested was good.

Reply:
So should we leave as it is?

Reply:
Actually, no, not quite.  I'm still not sure why we are catching the exception and rethrowing.  Wouldn't it be better to just not catch it at all?

Reply:
Only because of the warning. 🤔